File: buildy_lib/task_graph.py
"""Task dependency graph with topological sorting"""

import logging
from typing import Dict, List, Any

from .models import BuildTask

logger = logging.getLogger('buildy.task_graph')

class TaskGraph:
    """Task dependency graph with topological sorting"""

    def __init__(self):
        self.tasks: Dict[str, BuildTask] = {}
        self.execution_stages: List[List[str]] = []

    def add_task(self, task: BuildTask):
        """Add task to graph"""
        self.tasks[task.task_id] = task

    def build_execution_stages(self) -> List[List[str]]:
        """Build parallel execution stages using topological sort"""
        # Build dependency graph
        dependencies = {}
        dependents = {}

        for task_id, task in self.tasks.items():
            dependencies[task_id] = set(task.dependencies)
            dependents[task_id] = set()

        # Build reverse dependencies
        for task_id, deps in dependencies.items():
            for dep in deps:
                if dep in dependents:
                    dependents[dep].add(task_id)

        # Topological sort with parallel stages
        stages = []
        remaining_tasks = set(self.tasks.keys())

        while remaining_tasks:
            # Find tasks with no remaining dependencies
            ready_tasks = []
            for task_id in remaining_tasks:
                if not dependencies[task_id]:
                    ready_tasks.append(task_id)

            if not ready_tasks:
                # Circular dependency detected
                raise ValueError(f"Circular dependency detected in tasks: {remaining_tasks}")

            stages.append(ready_tasks)

            # Remove ready tasks and update dependencies
            for task_id in ready_tasks:
                remaining_tasks.remove(task_id)
                for dependent in dependents[task_id]:
                    dependencies[dependent].discard(task_id)

        self.execution_stages = stages
        return stages

    def get_execution_plan(self) -> Dict[str, Any]:
        """Get detailed execution plan"""
        if not self.execution_stages:
            self.build_execution_stages()

        total_time = 0
        max_parallel = 0
        total_cpu_hours = 0

        for stage in self.execution_stages:
            stage_time = 0
            stage_cpus = 0
            for task_id in stage:
                task = self.tasks[task_id]
                stage_time = max(stage_time, task.estimated_time)
                stage_cpus += task.resource_requirements.cpu_cores
                total_cpu_hours += (task.estimated_time / 3600) * task.resource_requirements.cpu_cores

            total_time += stage_time
            max_parallel = max(max_parallel, len(stage))

        return {
            'total_stages': len(self.execution_stages),
            'total_tasks': len(self.tasks),
            'estimated_time_seconds': total_time,
            'max_parallel_tasks': max_parallel,
            'total_cpu_hours': total_cpu_hours,
            'stages': [
                {f'stage_{i+1}': stage} 
                for i, stage in enumerate(self.execution_stages)
            ]
        }

File: goBuildy/task_graph.go
package main

import (
	"fmt"
	"log"
)

// TaskGraph represents a task dependency graph with topological sorting
type TaskGraph struct {
	Tasks           map[string]*BuildTask
	ExecutionStages [][]string
}

// NewTaskGraph creates a new TaskGraph
func NewTaskGraph() *TaskGraph {
	return &TaskGraph{
		Tasks:           make(map[string]*BuildTask),
		ExecutionStages: [][]string{},
	}
}

// AddTask adds a task to the graph
func (tg *TaskGraph) AddTask(task *BuildTask) {
	tg.Tasks[task.TaskID] = task
}

// BuildExecutionStages builds parallel execution stages using topological sort
func (tg *TaskGraph) BuildExecutionStages() ([][]string, error) {
	// Build dependency graph
	dependencies := make(map[string]map[string]bool)
	dependents := make(map[string]map[string]bool)

	for taskID, task := range tg.Tasks {
		dependencies[taskID] = make(map[string]bool)
		for _, dep := range task.Dependencies {
			dependencies[taskID][dep] = true
		}
		dependents[taskID] = make(map[string]bool)
	}

	// Build reverse dependencies
	for taskID, deps := range dependencies {
		for dep := range deps {
			if _, exists := dependents[dep]; exists {
				dependents[dep][taskID] = true
			}
		}
	}

	// Topological sort with parallel stages
	stages := [][]string{}
	remainingTasks := make(map[string]bool)
	for taskID := range tg.Tasks {
		remainingTasks[taskID] = true
	}

	for len(remainingTasks) > 0 {
		// Find tasks with no remaining dependencies
		readyTasks := []string{}
		for taskID := range remainingTasks {
			if len(dependencies[taskID]) == 0 {
				readyTasks = append(readyTasks, taskID)
			}
		}

		if len(readyTasks) == 0 {
			// Circular dependency detected
			remaining := []string{}
			for taskID := range remainingTasks {
				remaining = append(remaining, taskID)
			}
			return nil, fmt.Errorf("circular dependency detected in tasks: %v", remaining)
		}

		stages = append(stages, readyTasks)

		// Remove ready tasks and update dependencies
		for _, taskID := range readyTasks {
			delete(remainingTasks, taskID)
			for dependent := range dependents[taskID] {
				delete(dependencies[dependent], taskID)
			}
		}
	}

	tg.ExecutionStages = stages
	return stages, nil
}

// GetExecutionPlan gets detailed execution plan
func (tg *TaskGraph) GetExecutionPlan() (map[string]any, error) {
	if len(tg.ExecutionStages) == 0 {
		if _, err := tg.BuildExecutionStages(); err != nil {
			return nil, err
		}
	}

	totalTime := 0.0
	maxParallel := 0
	totalCPUHours := 0.0

	for _, stage := range tg.ExecutionStages {
		stageTime := 0.0
		stageCPUs := 0
		for _, taskID := range stage {
			task := tg.Tasks[taskID]
			if task.EstimatedTime > stageTime {
				stageTime = task.EstimatedTime
			}
			stageCPUs += task.ResourceRequirements.CPUCores
			totalCPUHours += (task.EstimatedTime / 3600.0) * float64(task.ResourceRequirements.CPUCores)
		}

		totalTime += stageTime
		if len(stage) > maxParallel {
			maxParallel = len(stage)
		}
	}

	stagesList := []map[string]any{}
	for i, stage := range tg.ExecutionStages {
		stagesList = append(stagesList, map[string]any{
			fmt.Sprintf("stage_%d", i+1): stage,
		})
	}

	return map[string]any{
		"total_stages":            len(tg.ExecutionStages),
		"total_tasks":             len(tg.Tasks),
		"estimated_time_seconds":  totalTime,
		"max_parallel_tasks":      maxParallel,
		"total_cpu_hours":         totalCPUHours,
		"stages":                  stagesList,
	}, nil
}

// LogExecutionPlan logs the execution plan
func (tg *TaskGraph) LogExecutionPlan() {
	plan, err := tg.GetExecutionPlan()
	if err != nil {
		log.Printf("ERROR: Failed to get execution plan: %v", err)
		return
	}

	log.Printf("Execution Plan:")
	log.Printf("  Total stages: %v", plan["total_stages"])
	log.Printf("  Total tasks: %v", plan["total_tasks"])
	log.Printf("  Estimated time: %.2f seconds", plan["estimated_time_seconds"])
	log.Printf("  Max parallel tasks: %v", plan["max_parallel_tasks"])
	log.Printf("  Total CPU hours: %.4f", plan["total_cpu_hours"])

	for i, stage := range tg.ExecutionStages {
		log.Printf("  Stage %d: %d tasks", i+1, len(stage))
	}
}

